File: app_mvp/src/main/java/com/example/jqorz/test1/mvp/delegate/fragment/FragmentDelegateImp.java
package com.example.jqorz.test1.mvp.delegate.fragment;


import com.example.jqorz.test1.mvp.MvpView;
import com.example.jqorz.test1.mvp.delegate.BaseDelegateCallback;
import com.example.jqorz.test1.mvp.delegate.MvpInternalDelegate;
import com.example.jqorz.test1.mvp.presenters.MvpPresenter;

/**
 * Created by jqorz on 2017/8/28.
 * 实现了FragmentDelegate，在生命周期里控制Presenter与MvpView：
 * 当Fragment执行到各个生命周期时，这里会执行对应的生命周期
 */

public class FragmentDelegateImp<P extends MvpPresenter, V extends MvpView> implements FragmentDelegate {
    private BaseDelegateCallback<P, V> basemvpDelegateCallback;
    private MvpInternalDelegate<P, V> mvpInternalDelegate;

    //传入BaseDelegateCallback 去控制Presenter
    public FragmentDelegateImp(BaseDelegateCallback<P, V> basemvpDelegateCallback) {
        if (basemvpDelegateCallback == null)
            throw new NullPointerException("the basemvpDelegateCallback in FragmentDelegate is null");
        this.basemvpDelegateCallback = basemvpDelegateCallback;
        mvpInternalDelegate = new MvpInternalDelegate<>(this.basemvpDelegateCallback);

    }



    @Override
    public void onCreateView() {
        mvpInternalDelegate.createPresenter();
        mvpInternalDelegate.attachView();
    }

    @Override
    public void onPause() {

    }

    @Override
    public void onResume() {

    }

    @Override
    public void onStop() {

    }

    @Override
    public void onDestroyView() {
        mvpInternalDelegate.detachView();
    }
}

File: app_mvp/src/main/java/com/example/jqorz/mvptest/lib_kotlin/delegate/fragment/FragmentDelegateImp.kt
package com.jqorz.tallybook.base.mvp.delegate.fragment

import com.jqorz.tallybook.base.mvp.MvpView
import com.jqorz.tallybook.base.mvp.delegate.BaseDelegateCallback
import com.jqorz.tallybook.base.mvp.delegate.MvpInternalDelegate
import com.jqorz.tallybook.base.mvp.presenters.MvpPresenter

/**
 * 实现了FragmentDelegate，在生命周期里控制Presenter与MvpView：
 * 当Fragment执行到各个生命周期时，这里会执行对应的生命周期
 */
class FragmentDelegateImp<P : MvpPresenter<V>?, V : MvpView>(baseMvpDelegateCallback: BaseDelegateCallback<P, V>?) : FragmentDelegate {
    private val mvpInternalDelegate: MvpInternalDelegate<P, V>
    override fun onCreateView() {
        mvpInternalDelegate.createPresenter()
        mvpInternalDelegate.attachView()
    }

    override fun onPause() {}
    override fun onResume() {}
    override fun onStop() {}
    override fun onDestroyView() {
        mvpInternalDelegate.detachView()
    }

    //传入BaseDelegateCallback 去控制Presenter
    init {
        if (baseMvpDelegateCallback == null) throw NullPointerException("the baseMvpDelegateCallback in FragmentDelegate is null")
        mvpInternalDelegate = MvpInternalDelegate(baseMvpDelegateCallback)
    }
}